spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -350,8 +355,24 @@ public void stop(Runnable callback) {
 				public void handle(Exception thrownException, final ConsumerRecord message) {
 					final byte[] key = message.key() != null ? Utils.toArray(ByteBuffer.wrap((byte[]) message.key()))
 							: null;
-					final byte[] payload = message.value() != null
+					byte[] payload = message.value() != null
 							? Utils.toArray(ByteBuffer.wrap((byte[]) message.value())) : null;
+					if (properties.getHeaderMode().equals(HeaderMode.embeddedHeaders)) {
+						EmbeddedHeadersMessageConverter embeddedHeadersMessageConverter = new EmbeddedHeadersMessageConverter();
+						try {
+							MessageValues messageValues = embeddedHeadersMessageConverter.extractHeaders(MessageBuilder.withPayload(payload).build(), true);
```

Comment:
Creation of `Message` can be avoided if we change the `EmbeddedHeaderMessageConverter` method signature like [this](https://github.com/ilayaperumalg/spring-cloud-stream/commit/6f22c99a11748a1a31b525c26af5b11589bfdd53)